### PR TITLE
expression: implement vectorized evaluation for `builtinCastDecimalAsDecimalSig`

### PR DESCRIPTION
**Pull request by catror**
PCP #12102

## What have you changed?

implement vectorized evaluation for `builtinCastDecimalAsDecimalSig`

## What is changed and how it works?
```bash
$$ go test -check.f TestVectorizedBuiltinCastFunc
PASS: builtin_cast_vec_test.go:130: testEvaluatorSuite.TestVectorizedBuiltinCastFunc    0.146s
OK: 1 passed
PASS
ok      github.com/pingcap/tidb/expression      0.166s

$$ go test -v -benchmem -bench=BenchmarkVectorizedBuiltinCastFunc -run=BenchmarkVectorizedBuiltinCastFunc -args "builtinCastDecimalAsDecimalSig"
goos: darwin
goarch: amd64
pkg: github.com/pingcap/tidb/expression
BenchmarkVectorizedBuiltinCastFunc/builtinCastDecimalAsDecimalSig-VecBuiltinFunc-8                 69488             16760 ns/op               0 B/op             0 allocs/op
BenchmarkVectorizedBuiltinCastFunc/builtinCastDecimalAsDecimalSig-NonVecBuiltinFunc-8              19612             61114 ns/op           40032 B/op           834 allocs/op
PASS
ok      github.com/pingcap/tidb/expression      3.200s

```

## Check List

Tests
  - Unit test

### CODE REVIEW

**Comment by catror** [comment]
@SunRunAway Sorry, I misunderstood something. It's fixed now, please take a look.

File: expression/builtin_cast_vec.go [hunk 1]
```diff
@@ -530,11 +530,32 @@ func (b *builtinCastRealAsTimeSig) vecEvalTime(input *chunk.Chunk, result *chunk
...
+
+	n := input.NumRows()
+	decs := result.Decimals()
+	sc := b.ctx.GetSessionVars().StmtCtx
+	for i := 0; i < n; i++ {
+		if result.IsNull(i) {
+			continue
+		}
+		dec := &types.MyDecimal{}
+		if !(b.inUnion && mysql.HasUnsignedFlag(b.tp.Flag) && decs[i].IsNegative()) {
```
Comment (SunRunAway) [inline expression/builtin_cast_vec.go]: We can cache the result of `b.inUnion && mysql.HasUnsignedFlag(b.tp.Flag)` outside the for loop.

Comment (catror) [reply expression/builtin_cast_vec.go]: > We can cache the result of `b.inUnion && mysql.HasUnsignedFlag(b.tp.Flag)` outside the for loop.

Done.

**Pull request merged**